components/user-mgt/org.wso2.carbon.identity.user.action/src/main/java/org/wso2/carbon/identity/user/action/internal/listener/ActionUserOperationEventListener.java

```diff
@@ -127,6 +127,61 @@ public boolean doPreUpdateCredentialByAdminWithID(String userID, Object credenti
         }
     }
 
+    /**
+     * This method is responsible for handling the pre add user action execution.
+     *
+     * @param userID           User id of the user.
+     * @param credential       Credential of the user.
+     * @param userStoreManager User store manager.
+     * @return True if the operation is successful.
+     * @throws UserStoreException If an error occurs while executing the action.
+     */
+    @Override
+    public boolean doPreAddUserWithID(String userID, Object credential, String[] roleList, Map<String, String> claims,
+                                      String profile, UserStoreManager userStoreManager) throws UserStoreException {
+
+        if (!isEnable()) {
+            return true;
+        }
+
+        if (!isPreAddUserWithIDExecutableFlow()) {
+            return true;
+        }
+
+        try {
+            UserActionContext userActionContext = new UserActionContext(
+                    new UserActionRequestDTO.Builder()
+                            .userId(userID)
+                            .password(getSecret(credential))
+                            .userStoreDomain(UserCoreUtil.getDomainName(userStoreManager.getRealmConfiguration()))
+                            .build());
+
+            ActionExecutionStatus<?> executionStatus =
+                    UserActionExecutorFactory.getUserActionExecutor(ActionType.PRE_UPDATE_PASSWORD)
+                            .execute(userActionContext,
+                                    IdentityContext.getThreadLocalCarbonContext().getTenantDomain());
+
+            if (executionStatus.getStatus() == ActionExecutionStatus.Status.SUCCESS) {
+                return true;
+            } else if (executionStatus.getStatus() == ActionExecutionStatus.Status.FAILED) {
+                Failure failure = (Failure) executionStatus.getResponse();
+                throw new UserActionExecutionClientException(
+                        UserActionError.PRE_UPDATE_PASSWORD_ACTION_EXECUTION_FAILED,
+                        failure.getFailureReason(), failure.getFailureDescription());
+            } else if (executionStatus.getStatus() == ActionExecutionStatus.Status.ERROR) {
+                Error error = (Error) executionStatus.getResponse();
+                throw new UserActionExecutionServerException(
+                        UserActionError.PRE_UPDATE_PASSWORD_ACTION_EXECUTION_ERROR,
+                        error.getErrorMessage(), error.getErrorDescription());
+            } else {
+                return false;
```

Comment:
Since this logic is identical in both scenarios(`doPreUpdateCredentialByAdminWithID()` and `doPreAddUserWithID()`), let's refactor it to a helper method.